parser/src/lexer/mod.rs

```diff
@@ -83,6 +83,14 @@ pub enum Token {
     #[token("$rand")]
     Rand,
 
+    /// A reserved keyword for accessing main columns by index
+    #[token("$main")]
+    MainIndex,
+
+    /// A reserved keyword for accessing aux columns by index
+    #[token("$aux")]
+    AuxIndex,
+
```

Comment:
1. I would move this up 2 sections and put it next to the existing main/aux keywords. I know that I specified in the task that these should only be accessible in transition constraint expressions, but I think that's no longer true. I think we should probably allow using them for boundary constraints as well now. 
2. I would rename these to `MainAccess` and `AuxAccess` to be more consistent with our internal terminology in the AST and IR.
3. I would rename the existing `Main` and `Aux` to `MainDecl` and `AuxDecl` so that there's more clarity between those keywords and these new keywords.